lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,469 @@
+      if (temporaryValue === null) {
+        temporaryValue = value;
+      } else {
+        temporaryValue = `${temporaryValue}, ${value}`;
+      }
+      if (!isNoCORSSafelistedRequestHeader(name, value)) {
+        return;
+      }
+    } else if (this.#guard === 'response' && forbiddenResponseHeaderNames.has(name)) {
+      return;
+    }
+
+    this.#headerList.append(name, value);
+
+    if (this.#guard === 'request-no-cors') {
+      this.#removePrivilegedNoCORSRequestHeaders();
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-headers-fill
+  #fill(init) {
+    if (typeof init !== 'object' || init === null) {
+      throw new ERR_INVALID_ARG_TYPE('init', 'object', init);
+    }
+    if (typeof init[SymbolIterator] === 'function') {
+      for (const header of init) {
+        if (header === null ||
+            typeof header[SymbolIterator] !== 'function' ||
+            typeof header === 'string') {
+          throw new ERR_INVALID_ARG_TYPE('init.header', 'Iterable', header);
+        }
+        const pair = ArrayFrom(header);
+        if (pair.length !== 2) {
+          throw new ERR_INVALID_ARG_TYPE('init.header', 'of length two', pair);
+        }
+        this.#append(header[0], header[1]);
+      }
+    } else {
+      ArrayPrototypeForEach(ObjectEntries(init), (header) => {
+        this.#append(header[0], header[1]);
+      })
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-headers-remove-privileged-no-cors-request-headers
+  #removePrivilegedNoCORSRequestHeaders() {
+    for (const headerName of privilegedNoCORSRequestHeaderNames) {
+      this.#headerList.delete(headerName);
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers
+  constructor(init) {
+    this.#headerList = new HeaderList();
+    this.#guard = 'none';
+    if (init !== undefined){
+      this.#fill(init);
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-append
+  append(name, value) {
+    this.#append(name, value);
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-delete
+  delete(name) {
+    name = normalizeName(name);
+    const lowerName = StringPrototypeToLowerCase(name);
+    if (this.#guard === 'immutable') {
+      throw new TypeError('this Headers object is immutable');
+    }
+    if (this.#guard === 'request' && forbiddenHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (this.#guard === 'request-no-cors' &&
+        !noCORSSafelistedRequestHeaderNames.has(lowerName) &&
+        !privilegedNoCORSRequestHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (this.#guard === 'response' &&
+        forbiddenResponseHeaderNames.has(lowerName)) {
+      return;
+    }
+    if (!this.#headerList.contains(lowerName)) {
+      return;
+    }
+    this.#headerList.delete(lowerName);
+    if (this.#guard === 'request-no-cors') {
+      this.#removePrivilegedNoCORSRequestHeaders();
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#dom-headers-get
+  get(name) {
+    name = normalizeName(name);
```

Comment:
Would it be better to normalize on header storage rather than on read?

Reply:
I called the function `normalizeName` to be consistent with `nomalizeValue` but it just does a check for:

> If name is not a name, then throw a TypeError.